### ISSUE TITLE
Obtaining the WindowProc of the User32.DefWindowProcW

### ISSUE THREAD

**Issue by httpdigest** [LWJGL/lwjgl3#127]
I am having tough times trying to create a simple window using the platform API. :)

At most I am struggling with how to get a `WindowProc` instance from the default window procedure function in User32.
I did not find an example in the lwjgl3 repository (I also looked at the 'offscreen-context' branch).

This code currently crashes the JVM:
```Java
long addr = GetProcAddress(GetModuleHandle("User32.dll"), "DefWindowProcW");
WindowProc proc = WindowProc.<WindowProc>create(addr);
```

The `addr` is not null, but something like `1993516028`. Is this the correct way of doing things? I was doing this because I needed an instance of WindowProc for the WNDCLASSEX.

Crash log:
```
siginfo: ExceptionCode=0xc0000005, reading address 0xffffffffffffffff

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  sun.misc.Unsafe.getAddress(J)J+0
j  org.lwjgl.system.MemoryAccess$MemoryAccessorUnsafe.memGetAddress(J)J+4
j  org.lwjgl.system.MemoryUtil.memGetAddress(J)J+4
j  org.lwjgl.system.libffi.Closure.create(J)Lorg/lwjgl/system/libffi/Closure;+17
j  org.lwjgl.demo.opengl.window.WindowWin32.main([Ljava/lang/String;)V+18
v  ~StubRoutines::call_stub
```

OS: Win7 x64

**Comment by Spasi** [comment]
The `WindowProc.create` call doesn't work because it's really `org.lwjgl.system.libffi.Closure.create`, which only works for libffi closures. You cannot use it with an arbitrary function address.

See the following code:

```java
private static long createDummyWindow(String className) {
	WNDCLASSEX wc = WNDCLASSEX.calloc();
	try {
		wc.cbSize(WNDCLASSEX.SIZEOF);
		wc.style(CS_HREDRAW | CS_VREDRAW);
		memPutAddress(wc.address() + WNDCLASSEX.LPFNWNDPROC, User32.getInstance().DefWindowProc);
		wc.hbrBackground(6); // COLOR_WINDOWFRAME
		wc.lpszClassName(className);

		if ( RegisterClassEx(wc) == 0 )
			throw new IllegalStateException("Failed to register dummy window class");
	} finally {
		nmemFree(memGetAddress(wc.address() + WNDCLASSEX.LPSZCLASSNAME));
		wc.free();
	}

	long hwnd = CreateWindowEx(
		0,
		className,
		"Dummy Window",
		WS_OVERLAPPEDWINDOW | WS_CLIPCHILDREN | WS_CLIPSIBLINGS,
		0, 0, 1, 1,
		NULL, NULL, NULL, NULL
	);

	if ( hwnd == NULL )
		throw new IllegalStateException("Failed to create dummy window");

	return hwnd;
}
```
There are two ugly things above that I may fix soon:

- The `lpfnWndProc` setter only accepts WindowProc instances. An overload that accepts any function address should be added.
- The `lpszClassName` setter uses encodes the string to `malloc`'ed memory, which is then exceptionally hard to free. The allocator should be parametric and it should be trivial to free the allocated memory.

**Comment by httpdigest** [comment]
Ah, okay. Thanks for the info and the example code!
I went another route now to implement a custom WindowProc class and then just delegate the invoke method to `User32.DefWindowProc(hwnd, uMsg, wParam, lParam)`

**Comment by httpdigest** [comment]
Hmm... no matter what I try, somehow CreateWindowEx will always error with 1407 -> Window class not found.
Do you know what's wrong with this code?
```Java
public class WindowWin32 extends WindowProc {
    public void run() {
        ByteBuffer className = MemoryUtil.memEncodeUTF16("MyUniqueWindowClass1234");
        ByteBuffer windowName = MemoryUtil.memEncodeUTF16("MyWindow");

        WNDCLASSEX wnd = WNDCLASSEX.malloc();
        wnd.cbSize(WNDCLASSEX.SIZEOF);
        wnd.hbrBackground(1);
        wnd.lpfnWndProc(this);
        wnd.lpszClassName(className);
        short err = User32.RegisterClassEx(wnd);
        if (err != 0) {
            throw new AssertionError("Could not register window class!");
        }

        long hwnd = User32.CreateWindowEx(0, className, windowName,
                User32.WS_OVERLAPPEDWINDOW | User32.WS_VISIBLE,
                0, 0, 800, 600, 0L, 0L, 0L, 0L);
        int lastError = WinBase.getLastError();
        if (lastError != 0) {
            throw new AssertionError("CreateWindowEx produced error: " + lastError);
        }
        System.err.println(hwnd);
    }
    public static void main(String[] args) {
        WindowWin32 main = new WindowWin32();
        main.run();
    }
    public long invoke(long hwnd, int uMsg, long wParam, long lParam) {
        return User32.DefWindowProc(hwnd, uMsg, wParam, lParam);
    }
}
```

**Comment by Spasi** [comment]
Two problems:

- You're using `malloc()` to allocate the WNDCLASSEX struct and you aren't setting all the fields. This might result in garbage values in the fields you don't set. Use `calloc()` instead.
- The return value of `RegisterClassEx` is an `ATOM`. An error is indicated by a 0 value, not != 0.

**Comment by httpdigest** [comment]
D'oh. :( Now it works. Thanks!

**Comment by Spasi** [comment]
The problems I mentioned above have been fixed. In the next nightly build the sample code becomes:

```java
private static long createDummyWindow(String className) {
	WNDCLASSEX wc = WNDCLASSEX.calloc();
	try {
		wc.cbSize(WNDCLASSEX.SIZEOF);
		wc.style(CS_HREDRAW | CS_VREDRAW);
		wc.lpfnWndProc(User32.getInstance().DefWindowProc);
		wc.hbrBackground(6); // COLOR_WINDOWFRAME
		wc.lpszClassName(className);

		if ( RegisterClassEx(wc) == 0 )
			throw new IllegalStateException("Failed to register dummy window class");
	} finally {
		wc.lpszClassNameFree();
		wc.free();
	}

	long hwnd = CreateWindowEx(
		0,
		className,
		"Dummy Window",
		WS_OVERLAPPEDWINDOW | WS_CLIPCHILDREN | WS_CLIPSIBLINGS,
		0, 0, 1, 1,
		NULL, NULL, NULL, NULL
	);

	if ( hwnd == NULL )
		throw new IllegalStateException("Failed to create dummy window");

	return hwnd;
}
```

**Comment by hakanai** [comment]
Sadly LWJGL now validates that hInstance has been set into the WNDCLASSEX structure, so that sample as currently written no longer works. Even though supposedly setting NULL for the hInstance is perfectly fine. :/

**Comment by Spasi** [comment]
@hakanai Thanks! It will be fixed in the next 3.3.2 snapshot.
